### ISSUE TITLE
[ FEATURE REQUEST ]  Option to add extra headers to entire collection

### ISSUE THREAD

**Issue by someshkoli** [postmanlabs/newman#2275]
1. Newman Version (can be found via `newman -v`):4.6.0
 2. OS details (type, version, and architecture): Linux Kali 5.3.0-kali2-amd64 #1 SMP Debian 5.3.9-3kali1 (2019-11-20) x86_64 GNU/Linux
 3. Are you using Newman as a library, or via the CLI? CLI
 3. Did you encounter this recently, or has this bug always been there:faced recently
 4. Expected behaviour:
 5. Command / script used to run Newman:
 6. Sample collection, and auxiliary files (minus the sensitive details):
 7. Screenshots (if applicable):

I was wondering do we have any option to add some extra header to  entire collection via command line ? say we want to change / replace the host or add some other header parameter to entire collection ??

**Comment by gittysachin** [comment]
@someshkoli right now we don't have any option to add extra headers to the entire collection. 

I'd like to work on this feature.

**Comment by someshkoli** [comment]
I have already worked on this and have the code ready with me :P.

**Comment by gittysachin** [comment]
Cool :) Then please go-ahead 👍

**Comment by shamasis** [comment]
One concern with this is that it might not be an accepted feature to be added in the Newman from product management standpoint. It’s wider purpose use cases has not been properly aligned with the Postman app.

**Comment by someshkoli** [comment]
Yeah true. I faced this issue and thought this could be a good feature because sometimes while working on a collection i need to add some common header to all the requests in the collection or edit existing header parameter for an entire collection. So this could be helpful if we have something like this through which we can specify some header or replace existing once globally across all the requests in the collection. This can also be adopted in postman app where you can set some properties for which are common for all the requests in the collection.

**Comment by shamasis** [comment]
@someshkoli ... bang on target. There has to be a way to layer common aspects to collections. But that shouldn’t be definition driven at this point, but should be sort of an operation driven approach using find and replace and other checks. One reason for us to double check is how do we handle debugability of this? With inherited scripts from folders, we are already seeing tonnes of support cases from our 10M users around people not realising that they have stuff in folder script. With folder or collection level header, this can become pronounced if the UX is not thought through. We are right now in the thinking phase of things for this. 

There’s also something to consider for non REST APIs ... such as GraphQL and other protocols we add in future. Ideas are welcome, but I’m sure it wouldn’t be a quick decision.

**Comment by someshkoli** [comment]
Yeah exactly. By saying header it becomes very specific, Which shouldn't be the case  from the program management point of view. Giving it a planned approach I think this could be a helpful feature.
Also i didn't quite understand what you said about the debug-ability of the feature ?

**Comment by shamasis** [comment]
Debugability - what I meant was, being able to quickly discover what actions on the system has resulted in a particular outcome. Say your request fails because of a faulty header that was inherited and it takes you two hours to figure out or even correlate that this was because of the inherited header. It would be a bad experience.

**Comment by someshkoli** [comment]
Yeah that is quite important and should be taken into consideration. Need to think a way to debug errors caused by the additional parameters keeping in mind the UX of the product.

**Comment by andrewmelnyk** [comment]
Bumping this thread because it is a feature I could use right now. Has anyone looked into it lately?
